### ISSUE TITLE
Import Item csv

### ISSUE THREAD

**Issue by nathanzky** [opensourcepos/opensourcepos#2744]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [] Reporting an issue of an unmodified OSPOS installation
- [] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 3.3.1 - c786d4
- OSPOS git commit hash is: 
- PHP version is: 7.0.21
- MySQL or MariaDB version is: 5.7.18
- OS and version is: OS: Linux rhel76 3.10.0-957.el7.x86_64
- WebServer is: Apache2
- Selected language is: US
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is:  LAMP

### Issue / Bug  / Question / New Feature
Bug 

Import Item csv is not working having this error when I click submit button
![error_Import_csv](https://user-images.githubusercontent.com/26484936/75678723-b38ede80-5ca7-11ea-8cb8-02f74151c442.png)

**Comment by andriux1990** [comment]
You must be doing something wrong ... because the CSV import works fine. check the columns that everything is in order and that each cell has what it should have, and if they have barcodes, make sure they are not already in the db

**Comment by daN4cat** [comment]
Have you generated the csv with excel or something else?

**Comment by myosotisgit** [comment]
@nathanzky 
I have spend a few hours on the importing of a csv file, generated from a very old version of ospos (v1.x). I finally managed.

Please note that ospos (current version) can be picky on what is imported! For example, 
- empty rows/columns will not work
- NULL values will not be imported, for example on the ID column
- any value in columns like 'allow alt descr' will enable the option. If you dont want the option enabled leave the column empty.

**Comment by myosotisgit** [comment]
Another tip, If you work with 1 or more stock locations use this procedure:

- First create the stock location(s) (config > Stock)
In my example i have 1 location called "winkel" so i changed the default value.

Go to items
download the csv template

Add your articles to this csv.
Upload the csv file.

I have included an xslc example (with my stock location column) that works

[Copy of import_items.xlsx](https://github.com/opensourcepos/opensourcepos/files/4278013/Copy.of.import_items.xlsx)

**Comment by daN4cat** [comment]
@objecttothis any chance you can improve the import feature to avoid the issues reported above?

**Comment by daN4cat** [comment]
@myosotisgit thanks a lot for all the tips, very good. I'll add to the FAQ this link.

**Comment by objecttothis** [comment]
> 
> 
> @objecttothis any chance you can improve the import feature to avoid the issues reported above?

Do you mean these?
> Please note that ospos (current version) can be picky on what is imported! For example,

>    empty rows/columns will not work
>    NULL values will not be imported, for example on the ID column
>    any value in columns like 'allow alt descr' will enable the option. If you dont want the option enabled leave the column empty.

I need more information on what "empty columns will not work" means.  If, by empty columns, it means that if any cell is empty then it kicks an error, then I can change that so that if required cells are empty it will kick an error, but otherwise will process the row without error, but I need clarification.  

I can change the code so that it just skips empty rows without error.

Generally the reason for the pickiness is that the CSV import has a powerful ability to mess up data, so the more controlled, the better.

I have plans to implement a feature to give you a csv of which rows failed and the option to import the remaining rows, so that users can get more specific information about what the problem was, as well as quickly fix problems and re-import the remaining.  This feature is not at the top of my list right now though.

The 'allow alt descr' and similar fields are textboxes, so it's by design that non-empty cells result in a checked box.  If it isn't already, I could modify the code so that NULL, '', ' ' (any number of spaces), 0 and FALSE all make the box unchecked.

If there are other things that I'm missing, please let me know.  I can work on all, but the detailed report/optional import successful rows now, but I need more details to begin.

**Comment by myosotisgit** [comment]
> I need more information on what "empty columns will not work" means. If, by empty columns, it means that if any cell is empty then it kicks an error, then I can change that so that if required cells are empty it will kick an error, but otherwise will process the row without error, but I need clarification.

Logically speaking, empty columns (no required ones) will throw an error. Makes sense.
But empty rows should be ignored. But you shouldnt have empty rows on the import file anyway i hope :-)

> Generally the reason for the pickiness is that the CSV import has a powerful ability to mess up data, so the more controlled, the better.

This is understood and appreciated!

> The 'allow alt descr' and similar fields are textboxes, so it's by design that non-empty cells result in a checked box. If it isn't already, I could modify the code so that NULL, '', ' ' (any number of spaces), 0 and FALSE all make the box unchecked.

With documentation this could be made clear :-)


> If there are other things that I'm missing, please let me know. I can work on all, but the detailed report/optional import successful rows now, but I need more details to begin.

Now that you mention it. The error modal (popup) disappears after a few seconds.
It would be nice if this modal would stay displayed until i click "remove" or X.

**Comment by daN4cat** [comment]
I agree for the unchecked cases.

**Comment by objecttothis** [comment]
I'll get on it then.  I have a bit of a backlog, so if it's needed before 3.4.0, I'll need some help.  Someone updating the documentation would speed things along.

Planned changes:
- Empty rows are ignored.
- If a checkbox value is 0, FALSE, '', or spaces then this will be submitted to the database as unchecked.  ANY other value will translate to checked.
- Update documentation to reflect the above two items.

The latter I have to build into #2739 anyway and empty rows shouldn't be too difficult to handle, so I might be able to get the above done sooner.

**Comment by compuls1v3** [comment]
I'm also getting the same error.

**Comment by objecttothis** [comment]
> I'm also getting the same error.

I think I can work on this Monday when I'm back in the office.

**Comment by myosotisgit** [comment]
I looked on other issues for the import csv functionality as i am looking to find an answer on these questions:
- Can i import with the item_id column ?
- Can i update, for example, stock quantities by re-importing an updated csv file?

Its seems that this functionality is planned for a future version?
See 
- https://github.com/opensourcepos/opensourcepos/issues/2636
- https://github.com/opensourcepos/opensourcepos/issues/1412
= https://github.com/opensourcepos/opensourcepos/issues/1469

**Comment by objecttothis** [comment]
@myosotisgit currently the import treats all rows as new.  There is a planned functionality to allow an update import, but several issues need to be navigated in the design of that feature.  You're welcome to contribute code if you would like to speed things along, Otherwise, it is a future planned change.

**Comment by objecttothis** [comment]
Pull Request #2739 changes behavior:
- Empty rows are ignored
- 0, FALSE values translate as an attribute checkbox being created but unchecked.
- "", NULL and spaces cause no checkbox attribute to be created.
- TRUE, 1 and all other values cause a checkbox attribute to be created and checked.
- Non-required fields no longer throw an error.

**Comment by objecttothis** [comment]
OK, the rest of this needs an owner

https://github.com/opensourcepos/opensourcepos/wiki/Import-data-from-CSV-file updated to reflect 
- the current file structure of the CSV Item import.
- The current handling of different attribute types (acceptable values, unacceptable values)

With the quarantine it is imperative that I get some of these other pull requests moving so that our business can switch to generating online sales and I can't deal with documentation now.

**Comment by objecttothis** [comment]
> 
> 
> I looked on other issues for the import csv functionality as i am looking to find an answer on these questions:
> 
>     * Can i import with the item_id column ?
> 
>     * Can i update, for example, stock quantities by re-importing an updated csv file?
> 
> 
> Its seems that this functionality is planned for a future version?
> See
> 
>     * #2636
> 
>     * #1412
> 
>     * #1469

@myosotisgit Please test #2821 as it provides CSV update functionality to the item import allowing you to provide the item_id and have it update the item rather than add a new item.  Please let me know if there are any bugs that you find.

**Comment by objecttothis** [comment]
@myosotisgit We are getting ready to merge #2821, but it would help if you could go to bintray and download the latest build labeled "Item_import_update" and test it out.  Tell me if you see any problems: https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=desc#files

**Comment by myosotisgit** [comment]
Sorry for the radio silence. I am dealing with a lot of work and some personal dealings Cannot say when this is over...

**Comment by objecttothis** [comment]
> 
> 
> Sorry for the radio silence. I am dealing with a lot of work and some personal dealings Cannot say when this is over...

Understood.  No worries.  Sorry for the added stress you're going through.

**Comment by hanisirfan** [comment]
I don't know if this is related or not but I'm also having a problem importing CSV data. The import was successful and the data is indeed in the database but it still say "No Items to display.".  I've install all the necessary stuff (PHP extensions etc). Oh btw this is hosted locally using XAMPP.

**Comment by objecttothis** [comment]
@hanisirfan what is the version hash that you are using currently?  You should see it in the footer of each page to the right of the version number or under Office -> Configuration -> Setup & Config

**Comment by hanisirfan** [comment]
OSPOS Installation Info: 3.3.2 - bb309c

It's actually working fine now but before I tried about 2-3 fresh install before. Fresh install means that I delete all the files and drop the database and redo the installation based on the INSTALL.md instruction.

Out of topic:
Can we actually see "Receivings" report just like the other reports? Maybe per day report? 

Thanks.

**Comment by nathanzky** [comment]
@hanisirfan I try to do fresh install and drop the database but Im still getting this error in item import. can you send me the steps on how you manage to do it and if you can share your items_csv format would be appreciated.

**Comment by hanisirfan** [comment]
@nathanzky I just follow this instructions.

(Local Install)
https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md

I actually did it couple of times for it to work, I don't know why it doesn't works the first time.

For the CSV file, I downloaded from the system after I installed it. If you still want it maybe later I can share with you. I believe I'm using the latest stable version isn't it @objecttothis

**Comment by objecttothis** [comment]
> 
> 
> OSPOS Installation Info: 3.3.2 - bb309c
> 
> It's actually working fine now but before I tried about 2-3 fresh install before. Fresh install means that I delete all the files and drop the database and redo the installation based on the INSTALL.md instruction.
> 
> Out of topic:
> Can we actually see "Receivings" report just like the other reports? Maybe per day report?
> 
> Thanks.

That build does not include my changes in the #2821 which have not yet been merged into the master because I'm waiting on code reviews by other contributors.  I recommend you download one of those builds from bintray and test that to see if it resolves the issue.

**Comment by hanisirfan** [comment]
For me, after doing couple of reinstall. It actually works and as I said before I don't know what could cause it. It's a bit weird because items doesn't show up in the view but actually exist in db. It's become more confusing after it will only show up in view after adding a sale for an item (only that item will be visible in items view but the rest isn't). I'm no expert, I could help find the cause of it if I actually are.

**Comment by objecttothis** [comment]
> 
> 
> @nathanzky I just follow this instructions.
> 
> (Local Install)
> https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md
> 
> I actually did it couple of times for it to work, I don't know why it doesn't works the first time.
> 
> For the CSV file, I downloaded from the system after I installed it. If you still want it maybe later I can share with you. I believe I'm using the latest stable version isn't it @objecttothis

@nathanzky the latest stable version does not have my changes and bug fixes from #2821 in it yet.

**Comment by daN4cat** [comment]
Now part of #3029.

**Comment by objecttothis** [comment]
> 
> 
> Now part of #3029.

Yep.  Once we get done with bugfix builds for 3.3.x it should be next in line to get merged into the master.
